bridge/jni/src/logger.rs

```diff
@@ -0,0 +1,223 @@
+use jni::{
+	objects::{GlobalRef, JClass, JObject, JString, JValue},
+	sys::jint,
+	JNIEnv, JavaVM,
+};
+use log;
+use std::{
+	panic::{catch_unwind, AssertUnwindSafe},
+	process::abort,
+};
+
+#[derive(Clone, Copy, Debug)]
+enum SLF4JLogLevel {
+	Trace = 0,
+	Debug = 10,
+	Info = 20,
+	Warn = 30,
+	Error = 40,
+}
+
+impl From<log::Level> for SLF4JLogLevel {
+	fn from(level: log::Level) -> Self {
+		use log::Level::*;
+		match level {
+			Error => Self::Error,
+			Warn => Self::Warn,
+			Info => Self::Info,
+			Debug => Self::Debug,
+			Trace => Self::Trace,
+		}
+	}
+}
+
+impl From<SLF4JLogLevel> for log::Level {
+	fn from(level: SLF4JLogLevel) -> Self {
+		use SLF4JLogLevel::*;
+		match level {
+			Error => Self::Error,
+			Warn => Self::Warn,
+			Info => Self::Info,
+			Debug => Self::Debug,
+			Trace => Self::Trace,
+		}
+	}
+}
+
+impl TryFrom<jint> for SLF4JLogLevel {
+	type Error = ();
+
+	fn try_from(level: jint) -> Result<Self, ()> {
+		match level {
+			0 => Ok(Self::Trace),
+			10 => Ok(Self::Debug),
+			20 => Ok(Self::Info),
+			30 => Ok(Self::Warn),
+			40 => Ok(Self::Error),
+			_ => Err(()),
+		}
+	}
+}
+
+impl From<SLF4JLogLevel> for &str {
+	fn from(level: SLF4JLogLevel) -> Self {
+		use SLF4JLogLevel::*;
+		match level {
+			Trace => "trace",
+			Debug => "debug",
+			Info => "info",
+			Warn => "warn",
+			Error => "error",
+		}
+	}
+}
+
+struct SLF4JLogger {
+	vm: JavaVM,
+	logger_class: GlobalRef,
+}
+
+impl SLF4JLogger {
+	fn new(env: JNIEnv, logger_obj: JObject) -> jni::errors::Result<Self> {
+		Ok(Self { vm: env.get_java_vm()?, logger_class: env.new_global_ref(logger_obj)? })
+	}
+
+	fn log_impl(&self, record: &log::Record) -> jni::errors::Result<()> {
+		let mut env = self.vm.attach_current_thread()?;
```

Comment:
Good catch. Captured from Signal, but we don't need it, since all of our code is initiated through the Java side, and the native side doesn't have any independent threads.